File: src/cn/vove7/plugin/rest/filetype/RestColorSettingsPage.java
package cn.vove7.plugin.rest.filetype;

import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighter;
import com.intellij.openapi.options.colors.AttributesDescriptor;
import com.intellij.openapi.options.colors.ColorDescriptor;
import com.intellij.openapi.options.colors.ColorSettingsPage;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Map;

/**
 * @author danblack
 */
public class RestColorSettingsPage implements ColorSettingsPage {

    private static final AttributesDescriptor[] DESCRIPTORS = new AttributesDescriptor[]{
            new AttributesDescriptor("Comment", RestHighlighter.COMMENT),
            new AttributesDescriptor("Method", RestHighlighter.METHOD),
            new AttributesDescriptor("Url", RestHighlighter.URL),
            new AttributesDescriptor("Header", RestHighlighter.HEADER),
            new AttributesDescriptor("Option", RestHighlighter.OPTION),
            new AttributesDescriptor("Param", RestHighlighter.PARAM),
            new AttributesDescriptor("Error", RestHighlighter.BAD_CHARACTER),
            new AttributesDescriptor("Separator", RestHighlighter.SEPARATOR),
            new AttributesDescriptor("Body", RestHighlighter.BODY),
    };

    @NotNull
    @Override
    public AttributesDescriptor[] getAttributeDescriptors() {
        return DESCRIPTORS;
    }

    @NotNull
    @Override
    public ColorDescriptor[] getColorDescriptors() {
        return ColorDescriptor.EMPTY_ARRAY;
    }

    @NotNull
    @Override
    public String getDisplayName() {
        return "Rest";
    }

    @Nullable
    @Override
    public Icon getIcon() {
        return null;
    }

    @NotNull
    @Override
    public SyntaxHighlighter getHighlighter() {
        return new RestHighlighter();
    }

    @NotNull
    @Override
    public String getDemoText() {
        return "# This is a comment\n" +
                "-- option = value\n" +
                "POST\n" +
                "http://site.com\n" +
                "&param=value\n" +
                "@Content-type: application/xml\n" +
                "<body>\n" +
                "This is a request body\n" +
                "</body>\n" +
                "%%%\n" +
                "@Content-length: 100\n" +
                "<response>\n" +
                "This is a response body\n" +
                "</response>";
    }

    @Nullable
    @Override
    public Map<String, TextAttributesKey> getAdditionalHighlightingTagToDescriptorMap() {
        return null;
    }
}

File: src/main/kotlin/cn/vove7/plugin/rest/filetype/RestColorSettingsPage.kt
package cn.vove7.plugin.rest.filetype

import com.intellij.openapi.options.colors.ColorSettingsPage
import com.intellij.openapi.options.colors.AttributesDescriptor
import cn.vove7.plugin.rest.filetype.RestColorSettingsPage
import com.intellij.openapi.options.colors.ColorDescriptor
import com.intellij.openapi.fileTypes.SyntaxHighlighter
import cn.vove7.plugin.rest.filetype.RestHighlighter
import com.intellij.openapi.editor.colors.TextAttributesKey
import javax.swing.Icon

/**
 * @author danblack
 */
class RestColorSettingsPage : ColorSettingsPage {
    override fun getAttributeDescriptors(): Array<AttributesDescriptor> {
        return DESCRIPTORS
    }

    override fun getColorDescriptors(): Array<ColorDescriptor> {
        return ColorDescriptor.EMPTY_ARRAY
    }

    override fun getDisplayName(): String {
        return "Rest"
    }

    override fun getIcon(): Icon? {
        return null
    }

    override fun getHighlighter(): SyntaxHighlighter {
        return RestHighlighter()
    }

    override fun getDemoText(): String {
        return """
            # This is a comment
            -- option = value
            POST
            http://site.com
            &param=value
            @Content-type: application/xml
            <body>
            This is a request body
            </body>
            %%%
            @Content-length: 100
            <response>
            This is a response body
            </response>
            """.trimIndent()
    }

    override fun getAdditionalHighlightingTagToDescriptorMap(): Map<String, TextAttributesKey>? {
        return null
    }

    companion object {
        private val DESCRIPTORS = arrayOf(
            AttributesDescriptor("Comment", RestHighlighter.COMMENT),
            AttributesDescriptor("Method", RestHighlighter.METHOD),
            AttributesDescriptor("Url", RestHighlighter.URL),
            AttributesDescriptor("Header", RestHighlighter.HEADER),
            AttributesDescriptor("Option", RestHighlighter.OPTION),
            AttributesDescriptor("Param", RestHighlighter.PARAM),
            AttributesDescriptor("Error", RestHighlighter.BAD_CHARACTER),
            AttributesDescriptor("Separator", RestHighlighter.SEPARATOR),
            AttributesDescriptor("Body", RestHighlighter.BODY)
        )
    }
}